pkg/autoupgrade/daemon.go

```diff
@@ -0,0 +1,379 @@
+							continue
+						}
+						app.Status.AvailableAppImage = imageKey.image
+						app.Status.ConfirmUpgradeAppImage = ""
+					case "notify":
+						if app.Status.ConfirmUpgradeAppImage == imageKey.image {
+							continue
+						}
+						app.Status.ConfirmUpgradeAppImage = imageKey.image
+						app.Status.AvailableAppImage = ""
+					default:
+						logrus.Warnf("Unrecognized auto-upgrade mode %v for %v", mode, app.Name)
+						continue
+					}
+					logrus.Infof("Triggering an auto-upprade of app %v because a new digest was detected for image %v",
+						appKey, imageKey.image)
+					if err := d.client.Status().Update(ctx, &app); err != nil {
+						logrus.Errorf("Problem updating %v: %v", appKey, err)
+						continue
+					}
+				}
+			}
+
+			// This app was checked on this run, so update the nextCheck time for this app
+			nextCheckTime, interval, err := calcNextCheck(defaultNextCheck, app)
+			if err != nil {
+				logrus.Errorf("Problem calculating next check time for app %v: %v", app.Name, err)
+				continue
+			}
+			d.appKeysToNextCheck[appKey] = nextCheckDetails{time: nextCheckTime, appSpecificInterval: interval}
+		}
+	}
+
+	nearestNextCheck := defaultNextCheck
+	for _, nextCheck := range d.appKeysToNextCheck {
+		if nextCheck.time.After(now) && nextCheck.time.Before(nearestNextCheck) {
+			nearestNextCheck = nextCheck.time
+		}
+	}
+
+	if defaultNextCheck.After(nearestNextCheck) {
+		go func() {
+			// Schedule the next sync for the next nearest interval
+			logrus.Debugf("Next auto-upgrade sync scheduled for: %v", nearestNextCheck)
+			time.Sleep(time.Until(nearestNextCheck))
+			Sync()
+		}()
+	}
+
+	return nil
+}
+
+func calcNextCheck(defaultNextCheck time.Time, app v1.AppInstance) (time.Time, string, error) {
+	if app.Spec.AutoUpgradeInterval != "" {
+		nextCheckInterval, err := time.ParseDuration(app.Spec.AutoUpgradeInterval)
+		if err != nil {
+			return time.Time{}, "", err
+		}
+		return time.Now().Add(nextCheckInterval), app.Spec.AutoUpgradeInterval, nil
+	}
+	return defaultNextCheck, "", nil
+}
+
+func removeTagPattern(image string) string {
+	p, ok := AutoUpgradePattern(image)
+	if !ok {
+		return image
+	}
+
+	return strings.TrimSuffix(image, ":"+p)
+}
+
+func AutoUpgradePattern(image string) (string, bool) {
+	// This first bit is adapted from https://github.com/google/go-containerregistry/blob/main/pkg/name/tag.go
+	// Split on ":"
+	parts := strings.Split(image, ":")
+	var tag string
+	// Verify that we aren't confusing a tag for a hostname w/ port for the purposes of weak validation.
+	if len(parts) > 1 && !strings.Contains(parts[len(parts)-1], "/") {
+		tag = parts[len(parts)-1]
+	}
+
+	return tag, strings.ContainsAny(tag, "#*")
+}
+
+func Mode(appSpec v1.AppInstanceSpec) (string, bool) {
+	_, isPat := AutoUpgradePattern(appSpec.Image)
+	on := (appSpec.AutoUpgrade != nil && *appSpec.AutoUpgrade) || (appSpec.NotifyUpgrade != nil && *appSpec.NotifyUpgrade) || isPat
```

Comment:
For nil bools feel free to add GetAutoUpgrade() method like the GetDevMode() method on appInstanceSpec.  Makes the code more readable.

Reply:
aye, done